File: PaleoRose/Classes/Layer Control/XRLayerLineArrow.m
//
//  XRLayerLineArrow.m
//  XRose
//
//  Created by Tom Moore on Sun Mar 14 2004.
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#import "XRLayerLineArrow.h"
#import "XRDataSet.h"
#import "XRStatistic.h"
#import "XRGeometryController.h"
#import "PaleoRose-Swift.h"
#import <Cocoa/Cocoa.h>
@implementation XRLayerLineArrow

-(id)initWithGeometryController:(XRGeometryController *)aController withSet:(XRDataSet *)aSet
{
    if (!(self = [super initWithGeometryController:aController])) return nil;
    if(self)
    {
        _theSet = aSet; //note: not retained by this object
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(statisticsDidChange:) name:XRDataSetChangedStatisticsNotification object:_theSet];
        if(aSet)
            [self setLayerName:[aSet name]];
        else
            [self setLayerName:@"unnamed"];
        [self setStrokeColor:[NSColor blackColor]];
        [self setFillColor:[NSColor blackColor]];

        _arrowSize = 1.0;
        _type = 0;
        _headType = 1;
        _showVector = YES;
        _showError = YES;
        [self generateGraphics];
    }
    return self;
}

-(id)initWithIsVisible:(BOOL)visible
                active:(BOOL)active
                 biDir:(BOOL)isBiDir
                  name:(NSString *)layerName
            lineWeight:(float)lineWeight
              maxCount:(int)maxCount
            maxPercent:(float)maxPercent
           strokeColor:(NSColor *)strokeColor
             fillColor:(NSColor *)fillColor
             arrowSize:(float)arrowSize
            vectorType:(int)vectorType
             arrowType:(int)arrowType
            showVector:(BOOL)showVector
             showError:(BOOL)showError
             datasetId:(int)datasetId {
    self = [super init];
    if (self) {
        _graphicalObjects = [[NSMutableArray alloc] init];
        _isVisible = visible;
        _isActive = active;
        _isBiDir = isBiDir;
        _layerName = layerName;
        _lineWeight = lineWeight;
        _maxCount = maxCount;
        _maxPercent = maxPercent;
        _strokeColor = strokeColor;
        _fillColor = fillColor;
        _arrowSize = arrowSize;
        _type = vectorType;
        _headType = arrowType;
        _showVector = showVector;
        _showError = showError;
        _datasetId = datasetId;
        _canFill = YES;
        _canStroke = YES;
        // Generate the color preview image for the table view
        [self resetColorImage];
    }
    return self;
}

-(void)setLayerName:(NSString *)layerName
{
    _layerName = [NSString stringWithFormat:@"Stat_%@",layerName];
    [[NSNotificationCenter defaultCenter] postNotificationName:XRLayerTableRequiresReload object:self];
    [[NSNotificationCenter defaultCenter] postNotificationName:XRLayerInspectorRequiresReload object:self];
}

-(void)generateGraphics
{
    if (!_theSet) {
        return;
    }

    if (!geometryController) {
        return;
    }

    NSMutableDictionary *theDictionary = [[NSMutableDictionary alloc] init];
    float vector = 0.0;
    float radius;
    float radiusCore = 0.0;
    float vectorStrength = 0.0;
    float errorAngle = 0.0;
    ArrowHead *anArrow;
    NSBezierPath *aPath;
    NSArray *stats = [_theSet currentStatistics];
    NSEnumerator *anEnum = [stats objectEnumerator];
    XRStatistic *theStat;
    [_graphicalObjects removeAllObjects];
    while(theStat = [anEnum nextObject])
    {
        if([theStat.statisticName isEqualToString:[NSString stringWithUTF8String:"θ̅"]])
            vector = [theStat floatValue];
        if([theStat.statisticName isEqualToString:[NSString stringWithUTF8String:"R̅"]])
            vectorStrength = [theStat floatValue];
        if([theStat.statisticName isEqualToString:[NSString stringWithUTF8String:"θ̅± (95%)"]])
            errorAngle = [theStat floatValue];
    }
    switch(_type)
    {

        case 1://vector strength
        {
            radiusCore = [geometryController radiusOfRelativePercent:0.0];
            radius = [geometryController radiusOfRelativePercent:vectorStrength];
        }
            break;
        case 2://outer ring simple vector
        {
            radiusCore = [geometryController unrestrictedRadiusOfRelativePercent:1.05];
            radius = radiusCore + 30.0;

        }
            break;
        default://simple vector
        {
            radiusCore = [geometryController radiusOfRelativePercent:0.0];
            radius = [geometryController radiusOfRelativePercent:1.0];

        }
            break;
    }
    aPath = [NSBezierPath bezierPath];
    [aPath moveToPoint:[geometryController rotationOfPoint:NSMakePoint(0.0,radiusCore) byAngle:vector]];
    [aPath lineToPoint:[geometryController rotationOfPoint:NSMakePoint(0.0,radius) byAngle:vector]];
    [theDictionary setObject:aPath forKey:@"vector"];

    anArrow = [[ArrowHead alloc] initWithSize:_arrowSize color:_strokeColor type:_headType];

    [anArrow positionAtLineEndpoint:[geometryController rotationOfPoint:NSMakePoint(0.0,radius) byAngle:vector] withAngle:vector];
    [theDictionary setObject:anArrow forKey:@"arrow"];
    [_graphicalObjects addObject:theDictionary];
    [self configureErrorWithVector:vector error:errorAngle];

    [[NSNotificationCenter defaultCenter] postNotificationName:XRLayerTableRequiresReload object:self];
    [[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
}

-(void)configureErrorWithVector:(float)vAngle error:(float)error
{
    NSBezierPath *errorPath = [NSBezierPath bezierPath];
    //NSLog(@"anglg %f error %f",vAngle,error);
    float radiusCore;
    float radius;
    float midRadius;
    float leftAngle = vAngle - error;
    float rightAngle = vAngle + error;
    if(vAngle > 180.0)
        return;
    if(leftAngle<0)
        leftAngle = 360+leftAngle;
    if(rightAngle >360)
        rightAngle = 360-rightAngle;
    radiusCore = [geometryController unrestrictedRadiusOfRelativePercent:1.05];
    radius = radiusCore+15;
    midRadius = radiusCore+7;
    [errorPath moveToPoint:[geometryController rotationOfPoint:NSMakePoint(0.0,radiusCore) byAngle:leftAngle]];
    [errorPath lineToPoint:[geometryController rotationOfPoint:NSMakePoint(0.0,radius) byAngle:leftAngle]];
    [errorPath moveToPoint:[geometryController rotationOfPoint:NSMakePoint(0.0,radiusCore) byAngle:rightAngle]];
    [errorPath lineToPoint:[geometryController rotationOfPoint:NSMakePoint(0.0,radius) byAngle:rightAngle]];

    [errorPath moveToPoint:[geometryController rotationOfPoint:NSMakePoint(0.0,midRadius) byAngle:rightAngle]];
    //NSLog(@"angles %f %f",leftAngle,rightAngle);
    rightAngle = 360 - rightAngle + 90;
    //NSLog(@"angles %f %f",leftAngle,rightAngle);
    while(rightAngle > 360.0)
        rightAngle = rightAngle-360.0;
    while(rightAngle < 0.00)
        rightAngle = 360.0+rightAngle;
    leftAngle = 360 - leftAngle + 90;
    //NSLog(@"angles %f %f",leftAngle,rightAngle);
    while(leftAngle < 0.0)
        leftAngle = 360.0+leftAngle;
    while(leftAngle > 360.0)
        leftAngle = leftAngle-360.0;
    //NSLog(@"*angles %f %f",leftAngle,rightAngle);
    [errorPath appendBezierPathWithArcWithCenter:NSMakePoint(0.0,0.0) radius:midRadius startAngle:rightAngle endAngle:leftAngle];
    if(errorPath)
        [_graphicalObjects addObject:errorPath];
}
-(void)drawRect:(NSRect)rect
{
    if (!_graphicalObjects || [_graphicalObjects count] == 0) {
        return;
    }

    [NSGraphicsContext saveGraphicsState];
    [_strokeColor set];

    if(_isVisible)
    {
        if(_showVector)
        {
            NSBezierPath *path;
            NSDictionary *dict = [_graphicalObjects objectAtIndex:0];

            if((path = [[_graphicalObjects objectAtIndex:0] objectForKey:@"vector"]))
            {
                [(NSBezierPath *)path setLineWidth:_lineWeight];
                [(NSBezierPath *)path stroke];
                [[dict objectForKey:@"arrow"] drawRect:rect];
            }
        }
        if((_showError)&&([_graphicalObjects count]>1))
        {
            NSBezierPath *path = [_graphicalObjects objectAtIndex:1];
            if(path)
            {
                [path setLineWidth:_lineWeight];
                [path stroke];
            }
        }
    }
    [NSGraphicsContext restoreGraphicsState];
}

-(void)setStrokeColor:(NSColor *)color
{
    _strokeColor = color;
    [self resetColorImage];
    //possibly post notification.
    for(int i=0;i<[_graphicalObjects count];i++)
    {
        if([[_graphicalObjects objectAtIndex:i] respondsToSelector:@selector(setColor:)])
            [[_graphicalObjects objectAtIndex:i] setColor:_strokeColor];
    }
    [[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
    [[NSNotificationCenter defaultCenter] postNotificationName:XRLayerInspectorRequiresReload object:self];

}

-(XRDataSet *)dataSet
{
    return _theSet;
}

-(void)statisticsDidChange:(NSNotification *)notification
{
    [self generateGraphics];
    [[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
    [[NSNotificationCenter defaultCenter] postNotificationName:XRLayerInspectorRequiresReload object:self];
}

-(void)setDataSet:(XRDataSet *)aSet
{
    _theSet = aSet;
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(statisticsDidChange:) name:XRDataSetChangedStatisticsNotification object:_theSet];
    [self generateGraphics];
}

-(NSString *)type
{
    return @"XRLayerLineArrow";
}

-(int)datasetId {
    // Return the stored dataset ID if we don't have a dataset reference yet
    // Otherwise return the actual dataset's ID
    if (_theSet) {
        return _theSet.setId;
    }
    return _datasetId;
}

-(float)arrowSize {
    return _arrowSize;
}

-(int)vectorType {
    return _type;
}

-(int)arrowType {
    return _headType;
}

-(BOOL)showVector {
    return _showVector;
}
-(BOOL)showError {
    return _showError;
}

@end

File: PaleoRose/Classes/Layer Control/XRLayerLineArrow.swift
// XRLayerLineArrow.swift
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Cocoa

@objc class XRLayerLineArrow: XRLayer {
    // MARK: - Properties

    private var theSet: XRDataSet?
    private var arrowSize: Float = 10.0
    private var type: Int = 0
    private var headType: Int = 0
    private var showVector: Bool = true
    private var showError: Bool = false

    // MARK: - Class Methods

    override class func classTag() -> String {
        "LineArrow"
    }

    // MARK: - Initialization

    @objc init(geometryController: XRGeometryController, withSet set: XRDataSet) {
        super.init(geometryController: geometryController)
        theSet = set
        setupDefaults()
    }

    @objc init(isVisible: Bool,
               active: Bool,
               biDir: Bool,
               name: String,
               lineWeight: Float,
               maxCount: Int,
               maxPercent: Float,
               strokeColor: NSColor,
               fillColor: NSColor,
               arrowSize: Float,
               vectorType: Int,
               arrowType: Int,
               showVector: Bool,
               showError: Bool)
    {
        super.init()

        setIsVisible(isVisible)
        setIsActive(active)
        setBiDirectional(biDir)
        setLayerName(name)
        setLineWeight(lineWeight)
        self.maxCount = maxCount
        self.maxPercent = maxPercent
        setStrokeColor(strokeColor)
        setFillColor(fillColor)

        self.arrowSize = arrowSize
        type = vectorType
        headType = arrowType
        self.showVector = showVector
        self.showError = showError
    }

    required init(geometryController: XRGeometryController) {
        super.init(geometryController: geometryController)
    }

    required init(geometryController: XRGeometryController, dictionary: [String: Any]) {
        super.init(geometryController: geometryController, dictionary: dictionary)
        configureSelf(with: dictionary)
    }

    override required init() {
        super.init()
        setupDefaults()
    }

    // MARK: - Setup Methods

    private func setupDefaults() {
        arrowSize = 10.0
        type = 0
        headType = 0
        showVector = true
        showError = false
    }

    private func configureSelf(with dictionary: [String: Any]) {
        if let size = dictionary["Arrow_Size"] as? Float {
            arrowSize = size
        }
        if let vectorType = dictionary["Vector_Type"] as? Int {
            type = vectorType
        }
        if let arrowType = dictionary["Arrow_Type"] as? Int {
            headType = arrowType
        }
        if let showVec = dictionary["Show_Vector"] as? String {
            showVector = (showVec == "YES")
        }
        if let showErr = dictionary["Show_Error"] as? String {
            showError = (showErr == "YES")
        }
    }

    // MARK: - Public Methods

    @objc func configureError(withVector vAngle: Float, error: Float) {
        // Configure error settings for vector and error angles
        generateGraphics()
        NotificationCenter.default.post(name: NSNotification.Name(XRLayerRequiresRedraw), object: self)
    }

    @objc func getDatasetId() -> Int {
        theSet?.datasetId ?? -1
    }

    @objc func getArrowSize() -> Float {
        arrowSize
    }

    @objc func getVectorType() -> Int {
        type
    }

    @objc func getArrowType() -> Int {
        headType
    }

    @objc func getShowVector() -> Bool {
        showVector
    }

    @objc func getShowError() -> Bool {
        showError
    }

    // MARK: - XML Support

    override func xmlTree(forVersion version: String) -> LITMXMLTree {
        let tree = super.baseXMLTree(forVersion: version)

        tree.addAttribute("Arrow_Size", value: String(format: "%.1f", arrowSize))
        tree.addAttribute("Vector_Type", value: String(type))
        tree.addAttribute("Arrow_Type", value: String(headType))
        tree.addAttribute("Show_Vector", value: showVector ? "YES" : "NO")
        tree.addAttribute("Show_Error", value: showError ? "YES" : "NO")

        if let datasetId = theSet?.datasetId {
            tree.addAttribute("Dataset_ID", value: String(datasetId))
        }

        return tree
    }

    override func configure(withXMLTree1_0 configureTree: LITMXMLTree) {
        super.configure(withXMLTree1_0: configureTree)

        if let attributes = configureTree.attributesDictionary() as? [String: String] {
            if let sizeStr = attributes["Arrow_Size"] {
                arrowSize = Float(sizeStr) ?? 10.0
            }
            if let typeStr = attributes["Vector_Type"] {
                type = Int(typeStr) ?? 0
            }
            if let arrowTypeStr = attributes["Arrow_Type"] {
                headType = Int(arrowTypeStr) ?? 0
            }
            showVector = attributes["Show_Vector"] == "YES"
            showError = attributes["Show_Error"] == "YES"
        }
    }

    // MARK: - Drawing

    override func generateGraphics() {
        // Generate line and arrow graphics based on settings
        if showVector {
            generateVectorGraphics()
        }
        if showError {
            generateErrorGraphics()
        }
    }

    private func generateVectorGraphics() {
        // Implement vector graphics generation
    }

    private func generateErrorGraphics() {
        // Implement error graphics generation
    }

    override func draw(_: NSRect) {
        // Draw line and arrow graphics
    }

    // MARK: - Notification Handlers

    override func geometryDidChange(_: Notification) {
        generateGraphics()
        NotificationCenter.default.post(name: NSNotification.Name(XRLayerRequiresRedraw), object: self)
    }
}
